stdlib/hash.gr

```diff
@@ -240,13 +239,18 @@ let rec hashOne = (val, depth) => {
  * A generic hash function that produces an integer from any value. If `a == b` then `Hash.hash(a) == Hash.hash(b)`.
  *
  * @param anything: The value to hash
+ * @param seed: The seed to use
  * @returns A hash for the given value
  *
+ * @example Hash.hash(a) == Hash.hash(a)
+ * @example Hash.hash(a, seed=1) == -1015171190
+ *
  * @since v0.1.0
+ * @history v0.6.0: Added ability to set the seed
  */
 @unsafe
-provide let hash = anything => {
-  h = seed
+provide let hash = (anything, seed=seed) => {
```

Comment:
I don't think we want to allow setting a seed on each hash. Instead, I think we want to add a submodule with a Hash record that has the seed attached and then create a singleton at the top-level module that passes it to the submodule `hash`. This way people can do `Hash.SeededHash.make()` and then `Hash.SeededHash.hash(hasher, anything)`

Reply:
I think this would be more elegant if `Hash.SeededHash.make()` returned. Function.

`Hash.SeededHash.make(seed: Number) -> ((val: a) -> Number)` thoughts?

Reply:
I don't hate it! Let's get some other feedback before doing the work.

Reply:
I like Jake's idea; however in that case I'm not sure it needs to be in a submodule if it's just one function; maybe could just make a `makeSeeded` function?

Reply:
I am good with that approach.

Reply:
In the case of the make function, we'd want seed to be defaulted and use that to construct our global hasher, right?

Reply:
I dont think we should make it defaulted I think the api should look like.
```
module Hash

provide let makeSeeded = (seed) => ((value: a) => Number)
let hash = makeSeeded(currentSeedGenCode)
provide let hash = (value) => hash
```
we could just directly provide `hash` but then I think we miss out on the for good graindoc we currently have. Thoughts?

Reply:
I don't actually think we can use the approach I suggested. It doesnt look like the grain type system supports it.
![image](https://github.com/grain-lang/grain/assets/40705786/0f5a42a1-1805-4831-8fbf-98e6efa97640)
I seem to be getting an error because it cannot generalize the param type in the hash function.

Reply:
Yeah, our type system can't really support that right now.

I don't love the idea of the top module and the seeded module having such different APIs. I think the `SeededHash` module would make more sense if we had functors, e.g. `SeededHash.Make()` which would give you just a module with a `hash` function on it.

It'd be breaking, but I'd like to see that same `hash(hasher, anything)` on both versions, with `make` and `makeSeeded` functions to make the hasher.